packages-tests/Caching/Cache/CacheTest.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\Tests\Caching\Cache;
+
+use Rector\Caching\Cache;
+use Rector\Caching\Enum\CacheKey;
+use Rector\Testing\PHPUnit\AbstractRectorTestCase;
+
+final class CacheTest extends AbstractRectorTestCase
+{
+    private Cache $cache;
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $this->cache = $this->getService(Cache::class);
+    }
+
+    protected function tearDown(): void
+    {
+        $this->cache->clear();
+    }
+
+    public function testHasFileChanged(): void
+    {
+        $this->cache->save('configuration_hash_srv-rector-php', CacheKey::CONFIGURATION_HASH_KEY, 'something');
+        $this->assertNotNull(
+            $this->cache->load('configuration_hash_srv-rector-php', CacheKey::CONFIGURATION_HASH_KEY)
+        );
+
+        $this->cache->save('fc9a36aa55dbb3d22e66e71d4f824cf3f8aba88d', CacheKey::CONFIGURATION_HASH_KEY, 'anything');
+        $this->cache->clean('fc9a36aa55dbb3d22e66e71d4f824cf3f8aba88d');
+
+        $this->assertNotNull( // clean actually also cleans other cache key
+            $this->cache->load('configuration_hash_srv-rector-php', CacheKey::CONFIGURATION_HASH_KEY)
```

Comment:
Not sure this assumption makes sense: when working with a cache you always need to be prepared for the situation that the cache cannot deliver the key you ask for.

in this example its a bit suprising, that because of impl details the key cannot be delivered but I think it might be a issue from the calling code that this gets a real problem.

therefore I asked for the whole picture, when this scenario actually creates a problem for you

Reply:
Ok, sorry. I can give real case I am experiencing. I have code base, where I scan about 5000 files and about 1% of them collide with `configuration_hash_srv-rector-php` which is cache key for rector's config file, if this key gets invalidated, no caching is applied so effectivelly caching is not working for us at all.

Reply:
Shot in the dark: As a workaround you could try changing

https://github.com/rectorphp/rector-src/blob/6238705e938a643cb84e0b4275cb682d26307975/packages/Caching/ValueObject/Storage/FileCacheStorage.php#L86-90

to

```
$this->smartFileSystem->remove([
            $cacheFilePaths->getFilePath(),
        ]);
```

this should fix the problem I guess, but will leave empty directories in Our cache folders after cleanup

Reply:
I can confirm this works.

Reply:
I dont know the details/background of the caching, just made a guess regarding the above fix/workaround. Maybe @TomasVotruba can tell

Reply:
@TomasVotruba So can we keep these folders?

Reply:
I have barely any knowledge about caching system, so I can't help here. It's based on implementation in PHPStan.

Reply:
@mssimi i sent a PR to sour fork which might fix the problem while preserving the old behavior

Reply:
@staabm Sorry, missed that. :)

Reply:
@staabm can we skip that phpstan error?

Reply:
@mssimi I intentionally used rmdir() instead of the filesystem adapter because we only want to delete the directory if it is empty. Did you verify the fix works for you?

@TomasVotruba do you agree adding these errors to the ignore list is acceptable?

Reply:
@staabm Yep I understand, but I don't know, if Tomas let us ignore that phpstan thing :) I can confirm it fixes my problem.